extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java

```diff
@@ -534,16 +542,139 @@ public void handle(RoutingContext event) {
             root = mainRouter;
         }
 
-        warnIfProxyAddressForwardingAllowedWithMultipleHeaders(httpConfiguration);
-        ForwardingProxyOptions forwardingProxyOptions = ForwardingProxyOptions.from(httpConfiguration);
-        if (forwardingProxyOptions.proxyAddressForwarding) {
+        if (httpConfiguration.proxy.proxyAddressForwarding) {
+            warnIfProxyAddressForwardingAllowedWithMultipleHeaders(httpConfiguration);
+            final ForwardingProxyOptions forwardingProxyOptions = ForwardingProxyOptions.from(httpConfiguration);
+            final TrustedXForwardedProxyCheckBuilder proxyCheckBuilder = forwardingProxyOptions.trustedXForwardedProxyCheckBuilder;
             Handler<HttpServerRequest> delegate = root;
-            root = new Handler<HttpServerRequest>() {
-                @Override
-                public void handle(HttpServerRequest event) {
-                    delegate.handle(new ForwardedServerRequestWrapper(event, forwardingProxyOptions));
-                }
-            };
+            if (proxyCheckBuilder == null) {
+                // no proxy check => we do not restrict who can send `X-Forwarded` or `X-Forwarded-*` headers
+                final TrustedXForwardedProxyCheck allowAllProxyCheck = allowAll();
+                root = new Handler<HttpServerRequest>() {
+                    @Override
+                    public void handle(HttpServerRequest event) {
+                        delegate.handle(new ForwardedServerRequestWrapper(event, forwardingProxyOptions, allowAllProxyCheck));
+                    }
+                };
+            } else {
+                // restrict who can send `X-Forwarded` or `X-Forwarded-*` headers
+                root = new Handler<HttpServerRequest>() {
```

Comment:
This is complex enough that it should really be a top level class.

Reply:
Done, thx.